src/formats/PDB.cpp

```diff
@@ -670,101 +774,76 @@ void PDBFormat::write_next(const Frame& frame) {
-                        "residue '{}' chain id is too long, it will be truncated",
-                        chainid
-                    );
-                    chainid = chainid[0];
-                }
-            } else {
-                chainid = "X";
-            }
+        assert(r.resname.length() <= 3);
 
-            if (residue->get("insertion_code") &&
-                residue->get("insertion_code")->kind() == Property::STRING) {
-                inscode = residue->get("insertion_code")->as_string();
-                if (inscode.length() > 1) {
-                    warning("PDB writer",
-                        "residue '{}' insertion code is too long, it will be truncated",
-                        inscode
-                    );
-                    inscode = inscode[0];
-                }
-            } else {
-                inscode = " ";
-            }
-        } else {
-            resname = "XXX";
-            chainid = "X";
-            auto value = max_resid++;
-            if (value < 9999) {
-                resid = to_pdb_index(value);
-            } else {
-                resid = "  -1";
-            }
+        if (last_residue && last_residue->chainid != r.chainid && needs_ter_record(*last_residue)) {
+            file_.print("TER   {: >5}      {:3} {:1}{: >4s}{:1}\n",
+                to_pdb_index(i + ter_count),
+                last_residue->resname, last_residue->chainid, last_residue->resid, last_residue->inscode);
+            ter_serial_numbers.push_back(i + ter_count);
+            ++ter_count;
         }
 
-        assert(resname.length() <= 3);
         auto& pos = positions[i];
         check_values_size(pos, 8, "atomic position");
-
         file_.print(
             "{: <6}{: >5} {: <4s}{:1}{:3} {:1}{: >4s}{:1}   {:8.3f}{:8.3f}{:8.3f}{:6.2f}{:6.2f}          {: >2s}\n",
-            atom_hetatm, to_pdb_index(i), frame[i].name(), altloc, resname, chainid, resid, inscode, pos[0], pos[1], pos[2], 1.0, 0.0, frame[i].type()
+            r.atom_hetatm, to_pdb_index(i + ter_count), frame[i].name(), altloc,
+            r.resname, r.chainid, r.resid, r.inscode,
+            pos[0], pos[1], pos[2], 1.0, 0.0, frame[i].type()
         );
+
+        if (residue) {
+            last_residue = std::move(r);
+        }
+        else {
+            last_residue = nullopt;
+        }
+
     }
 
     auto connect = std::vector<std::vector<size_t>>(frame.size());
     for (auto& bond : frame.topology().bonds()) {
+        if (is_atom_record[bond[0]] && is_atom_record[bond[1]]) { // both must be standard residue atoms
+            continue;
+        }
         if (bond[0] > 99999 || bond[1] > 99999) {
             warning("PDB writer", "atomic index is too big for CONNECT, removing the bond between {} and {}", bond[0], bond[1]);
             continue;
         }
-        connect[bond[0]].push_back(bond[1]);
-        connect[bond[1]].push_back(bond[0]);
+
+        auto lower_b0 = std::lower_bound(ter_serial_numbers.begin(),
```

Comment:
This needs a comment explaining what is going on =)

Reply:
Sure thing!